### PR TITLE
swap fee now reinvested back into pool

### LINKED ISSUE #314: fix: remove swap fee rebate in ALEX and re-invest fee rebate into pool
With the introduction of AMM-based price oracle, swap fee rebate in ALEX creates a circular reference. So this needs to be handled (if we want) by a separate contract (so pools send rebate to the new contract (multisig?) which then converts to ALEX before paying out).

In absence of fee rebate in ALEX, the accrued swap fee should be re-invested into pool, so when liquidity is added or removed, the swap fee accrued prior to adding and until removal is properly handled.

### PR DESCRIPTION
**Pull request by fiftyeightandeight**
closes #314.
Need to update test cases.

### CODE REVIEW

**Comment by hyossid** [comment]
I see something weird.

you mentioned that fee is decomposed to fee-rebate and remainder goes to vault ( fee = fee-rebate + remainder)

but I don't see the remainder part sent to vault. 

for example on fixed weight pool, I think there should be a term like (remainder (- fee fee-rebate)), where remainder is sent to vault.

Please let me know if I got something wrong

**Comment by fiftyeightandeight** [comment]
You see that, for swap-x-for-y, dx moves from sender's wallet to vault, but the sender receives dy based on dx-net-fees. So vault receives dx, but returns dy net of fees. At the same time, balance-x (ie amount of token-x due to LPs by vault) increases by dx-net-fees + fee-rebate, so net net vault balance increases by (fee - fee-rebate), while LPs earn fee-rebate and sender receives dy against dx net of fees.

File: clarity/contracts/pool/collateral-rebalancing-pool.clar [hunk 1]
```diff
@@ -579,16 +582,16 @@
...
+                (fee (mul-up dx (get fee-rate-x pool)))
+                (fee-rebate (mul-down fee (get fee-rebate pool)))
                 (dx-net-fees (if (<= dx fee) u0 (- dx fee)))
                 (dy (try! (get-y-given-x token collateral expiry dx-net-fees)))
 
                 (pool-updated
                     (merge pool
                         {
-                            balance-x: (+ balance-x dx-net-fees),
+                            balance-x: (+ (+ balance-x dx-net-fees) fee-rebate),
```
Comment (LNow) [inline clarity/contracts/pool/collateral-rebalancing-pool.clar:593]:
```suggestion
                            balance-x: (+ balance-x dx-net-fees fee-rebate),
```
Clarity addition accepts 2 or more arguments, therefore this one can be coded as only one operation.

File: clarity/contracts/pool/collateral-rebalancing-pool.clar [hunk 2]
```diff
@@ -632,16 +634,16 @@
...
                 (dy-net-fees (if (<= dy fee) u0 (- dy fee)))
                 (dx (try! (get-x-given-y token collateral expiry dy-net-fees)))        
 
                 (pool-updated
                     (merge pool
                         {
                             balance-x: (if (<= balance-x dx) u0 (- balance-x dx)),
-                            balance-y: (+ balance-y dy-net-fees),                      
-                            fee-balance-y: (+ (get fee-balance-y pool) fee),
+                            balance-y: (+ (+ balance-y dy-net-fees) fee-rebate),
```
Comment (LNow) [inline clarity/contracts/pool/collateral-rebalancing-pool.clar:646]:
```suggestion
                            balance-y: (+ balance-y dy-net-fees fee-rebate),
```

File: clarity/contracts/pool/fixed-weight-pool.clar [hunk 3]
```diff
@@ -306,20 +297,19 @@
...
                 (dx-net-fees (if (<= dx fee) u0 (- dx fee)))
+                (fee-rebate (mul-down fee (get fee-rebate pool)))
     
                 (dy (try! (get-y-given-x token-x-trait token-y-trait weight-x weight-y dx-net-fees)))
 
                 (pool-updated
                     (merge pool
                         {
-                        balance-x: (+ balance-x dx-net-fees),
+                        balance-x: (+ (+ balance-x dx-net-fees) fee-rebate),
```
Comment (LNow) [inline clarity/contracts/pool/fixed-weight-pool.clar:311]:
```suggestion
                        balance-x: (+ balance-x dx-net-fees fee-rebate),
```

File: clarity/contracts/pool/fixed-weight-pool.clar [hunk 4]
```diff
@@ -352,21 +342,19 @@
...
 
                 (dx (try! (get-x-given-y token-x-trait token-y-trait weight-x weight-y dy-net-fees)))
 
                 (pool-updated
                     (merge pool
                         {
                         balance-x: (if (<= balance-x dx) u0 (- balance-x dx)),
-                        balance-y: (+ balance-y dy-net-fees),
-                        fee-balance-y: (+ fee fee-balance-y),
+                        balance-y: (+ (+ balance-y dy-net-fees) fee-rebate),
```
Comment (LNow) [inline clarity/contracts/pool/fixed-weight-pool.clar:357]:
```suggestion
                        balance-y: (+ balance-y dy-net-fees fee-rebate),
```

File: clarity/contracts/pool/yield-token-pool.clar [hunk 5]
```diff
@@ -300,26 +297,24 @@
...
 
                 (dy (try! (get-y-given-x the-aytoken dx-net-fees)))
 
                 (pool-updated
                     (merge pool
                         {
-                            balance-token: (+ balance-token dx-net-fees),
-                            balance-aytoken: (if (<= balance-aytoken dy) u0 (- balance-aytoken dy)),
-                            fee-balance-token: (+ fee-balance-token fee)
+                            balance-token: (+ (+ balance-token dx-net-fees) fee-rebate),
```
Comment (LNow) [inline clarity/contracts/pool/yield-token-pool.clar:316]:
```suggestion
                            balance-token: (+ balance-token dx-net-fees fee-rebate),
```

File: clarity/contracts/pool/yield-token-pool.clar [hunk 6]
```diff
@@ -345,25 +340,24 @@
...
+
                 (dx (try! (get-x-given-y the-aytoken dy-net-fees)))
 
                 (pool-updated
                     (merge pool
                         {
                             balance-token: (if (<= balance-token dx) u0 (- balance-token dx)),
-                            balance-aytoken: (+ balance-aytoken dy-net-fees),
-                            fee-balance-aytoken: (+ fee-balance-aytoken fee)
+                            balance-aytoken: (+ (+ balance-aytoken dy-net-fees) fee-rebate)
```
Comment (LNow) [inline clarity/contracts/pool/yield-token-pool.clar:360]:
```suggestion
                            balance-aytoken: (+ balance-aytoken dy-net-fees fee-rebate)
```

**Pull request merged**